test/optimizedForceMove.test.ts

```diff
@@ -0,0 +1,490 @@
+    whoSignedWhat = [0, 0, 2];
+    nStates = 3;
+    for (largestTurnNum = 2; largestTurnNum < 14; largestTurnNum += nParticipants) {
+      expect(
+        await optimizedForceMove.acceptableWhoSignedWhat(
+          whoSignedWhat,
+          largestTurnNum,
+          nParticipants,
+          nStates,
+        ),
+      ).toBe(false);
+    }
+  });
+});
+
+describe('_recoverSigner', () => {
+  // following https://docs.ethers.io/ethers.js/html/cookbook-signing.html
+  const privateKey = '0x0123456789012345678901234567890123456789012345678901234567890123';
+  const wallet = new ethers.Wallet(privateKey);
+  const msgHash = ethers.utils.id('Hello World');
+  const msgHashBytes = arrayify(msgHash);
+  it('recovers the signer correctly', async () => {
+    const sig = await sign(wallet, msgHashBytes);
+    expect(await optimizedForceMove.recoverSigner(msgHash, sig.v, sig.r, sig.s)).toEqual(
+      wallet.address,
+    );
+  });
+});
+
+describe('_validSignatures', () => {
+  const participants = [];
+  let stateHash;
+  const stateHashes = [];
+  let wallet;
+  let sig;
+  const sigs = [];
+  let brokenSigs;
+  const whoSignedWhat = [];
+  const largestTurnNum = 2;
+  it('returns true (false) for a correct (incorrect) set of signatures on n states', async () => {
+    for (let i = 0; i < 3; i++) {
+      wallet = ethers.Wallet.createRandom();
+      participants[i] = wallet.address;
+      stateHash = ethers.utils.id('Commitment' + i);
+      stateHashes[i] = stateHash;
+      sig = await sign(wallet, stateHash);
+      sigs[i] = {v: sig.v, r: sig.r, s: sig.s};
+      whoSignedWhat[i] = i;
+    }
+    expect(
+      await optimizedForceMove.validSignatures(
+        largestTurnNum,
+        participants,
+        stateHashes,
+        sigs,
+        whoSignedWhat,
+      ),
+    ).toBe(true);
+    brokenSigs = sigs.reverse();
+    expect(
+      await optimizedForceMove.validSignatures(
+        largestTurnNum,
+        participants,
+        stateHashes,
+        brokenSigs,
+        whoSignedWhat,
+      ),
+    ).toBe(false);
+  });
+  it('returns true (false) for a correct (incorrect) set of signatures on 1 state', async () => {
+    stateHash = ethers.utils.id('Commitment' + largestTurnNum);
+    for (let i = 0; i < 3; i++) {
+      wallet = ethers.Wallet.createRandom();
+      participants[i] = wallet.address;
+      sig = await sign(wallet, stateHash);
+      sigs[i] = {v: sig.v, r: sig.r, s: sig.s};
+      whoSignedWhat[i] = 0;
+    }
+    expect(
+      await optimizedForceMove.validSignatures(
+        largestTurnNum,
+        participants,
+        [stateHash],
+        sigs,
+        whoSignedWhat,
+      ),
+    ).toBe(true);
+    brokenSigs = sigs.reverse();
+    expect(
+      await optimizedForceMove.validSignatures(
+        largestTurnNum,
+        participants,
+        [stateHash],
+        brokenSigs,
+        whoSignedWhat,
+      ),
+    ).toBe(false);
+  });
+});
+
+describe('forceMove', () => {
```

Comment:
Should probably ultimately have checks like:
1. It accepts a forceMove for an open channel (first challenge)
2. It accepts a forceMove for an open channel (subsequent challenge, higher turnNum)
3. It rejects a forceMove for an open channel if the turnNum is too small (subsequent challenge, turnNumRecord would decrease)
3. It rejects a forceMove when a challenge is already underway
4. It rejects a forceMove for a finalized channel
5. It rejects a forceMove with an incorrect challengerSig
6. It rejects a forceMove with the states don't form a validTransition chain
7. It rejects a forceMove when one state isn't correctly signed

It might be worth adding these as TODOs. I don't think they need to be done here and now.

Reply:
TODO added